src/shared/signature.cpp

```diff
@@ -1,58 +1,83 @@
 // ./a.out < xochitl
 // ./a.out < remarkable-shutdown
 
-#include <string.h>
-#include <sys/stat.h>
-#include <sys/types.h>
-#include <unistd.h>
-#include <errno.h>
+#include <string>
+#include <fstream>
+#include <sstream>
+
 
 namespace swtfb {
-// from https://stackoverflow.com/a/14002993/442652
-char *read_file(const char *filename, int *out_size) {
-  FILE *f = fopen(filename, "rb");
-  if (f == NULL) {
-    fprintf(stderr, "Unable to open %s: %s\n", filename, strerror(errno));
-    exit(-1);
+struct Signature {
+  bool indirect;
+  std::string bytes;
+  int offset;
+};
+
+std::string read_file(const std::string& path) {
+  std::ostringstream output;
+  std::ifstream input_file{path, std::ifstream::binary};
+  output << input_file.rdbuf();
+  return std::string(output.str());
+}
+
+std::int32_t decode_jump_offset(std::string instruction) {
+  // Assuming A1 or A2 (4 bytes) instruction encoding
+  // See section A8.8.25 of the ARMv7 Architecture Reference Manual
+  char byte_0 = instruction[0];
+  char byte_1 = instruction[1];
+  char byte_2 = instruction[2];
+  char byte_3 = instruction[3];
+
+  if (((byte_3 >> 1) & 0b111) != 0b101) {
+    // Not a jump instruction
+    return 0;
   }
-  fseek(f, 0, SEEK_END);
-  long fsize = ftell(f);
-  fseek(f, 0, SEEK_SET); /* same as rewind(f); */
 
-  char *string = (char *)malloc(fsize + 1);
-  fread(string, 1, fsize, f);
-  fclose(f);
+  std::int32_t res = (
+    (byte_2 << 16)
+    | (byte_1 << 8)
+    | byte_0
+  );
+
+  bool is_negative = byte_2 >> 7;
 
-  string[fsize] = 0;
-  *out_size = fsize;
+  if (is_negative) {
+    res |= (0xff << 24);
+  }
 
-  return string;
+  return (res << 2) + 8;
 }
 
-char *locate_signature(const char *path, const char *find, int N) {
-  int size;
-  char *data = read_file(path, &size);
-
-  int offset = 0x10000;
-  char *ret = NULL;
-
-  bool found;
-  for (int i = 0; i < size - N; i++) {
-    found = true;
-    for (int j = 0; j < N; j++) {
-      if (find[j] != data[i + j]) {
-        found = false;
-        break;
-      }
+void *locate_signature(const std::string& data, const Signature& signature) {
+  const size_t base_offset = 0x10000;
+  const size_t position = data.find(signature.bytes);
```

Comment:
how fast is data.find on the xochitl's binary source?

Reply:
I haven’t benchmarked it, but it’s ([at least in libstdc++](https://github.com/gcc-mirror/gcc/blob/bc046a60cfdd7145fd1e644184ced04d89feb871/libstdc%2B%2B-v3/include/bits/basic_string.tcc#L1205-L1239)) essentially using the same O( _N_ × _M_ ) algorithm as was manually implemented here before. It’s potentially slightly faster since it’s using `memcmp` instead of comparing each byte in a loop.

Reply:
maybe faster, maybe not